### ISSUE TITLE
Warning "root" does not have permission to access the dev dir 

### ISSUE THREAD

**Issue by tjwebb** [nodejs/node-gyp#454]
```
gyp WARN EACCES user "root" does not have permission to access the dev dir "/home/travis/.node-gyp/0.10.28"
```
I'm filing this here because `node-gyp` is the only package that seems to throw this issue. My project depends on some modules which build with `node-gyp`, and I see this warning for every one of them during 'npm install'. Does anyone have insight on what this warning means?

**Comment by fooooorrest** [comment]
This error occurred for me when trying to install player and other libraries that depended on speaker.  
I got it to work by first installing libasound2-dev with:
`$ sudo apt-get install libasound2-dev`
Then installed speaker with :
`sudo npm install --unsafe-perm --verbose -g speaker`
and finally installed player with
`sudo npm install --unsafe-perm --verbose -g player`

looks good from the console output, but i haven't tried to play a file yet.

**Comment by apollolm** [comment]
FWIW - `sudo npm install --unsafe-perm` solved this problem for me as well on Ubuntu 14.04, Node 0.10.24.

**Comment by jordan112** [comment]
Yeah, I saw that.  Tried a bunch of stuff from it, but no dice.  I noticed that it does it even if I try to install bcrypt.  I'm going to close this issue because clearly its not a bug with Strider itself.

    npm install bcrypt -g

**Comment by jyoungrok** [comment]
After removing /root/.node-gyp, sudo npm install --unsafe-perm --verbose -g solved this problem

**Comment by donaldpipowitch** [comment]
I have a similar problem reported here: https://github.com/npm/npm/issues/7764. Is this a problem with npm or node-gyp? No solution solved my problem. When I use `--unsafe-perm` I get rid of all error messages, but the installation process runs endlessly.

**Comment by cliveb** [comment]
At node school,  same issue.
$ sudo npm install -g stream-adventure 
npm-debug-log

0 info it worked if it ends with ok
1 verbose cli [ 'node',
1 verbose cli '/usr/local/bin/npm',
1 verbose cli 'install',
1 verbose cli '--unsafe-perm',
1 verbose cli '--verbose',
1 verbose cli '-g' ]
2 info using npm@2.7.4
3 info using node@v0.12.2
4 silly cache add args [ '.', null ]
5 verbose cache add spec .
6 silly cache add parsed spec { raw: '.',
6 silly cache add scope: null,
6 silly cache add name: null,
6 silly cache add rawSpec: '.',
6 silly cache add spec: '/Users/seebe',
6 silly cache add type: 'local' }
7 error addLocal Could not install /Users/seebe
8 verbose stack Error: EISDIR, read
8 verbose stack at Error (native)
9 verbose cwd /Users/seebe
10 error Darwin 13.4.0
11 error argv "node" "/usr/local/bin/npm" "install" "--unsafe-perm" "--verbose" "-g"
12 error node v0.12.2
13 error npm v2.7.4
14 error code EISDIR
15 error errno -21
16 error EISDIR, read
17 error If you need help, you may report this error at:
17 error https://github.com/npm/npm/issues
18 verbose exit [ -21, true ]

**Comment by ranisalt** [comment]
--unsafe-perm also solved for me, installing on a Gentoo server

**Comment by ogrotten** [comment]
fyi about `--unsafe-perm`, I found ONE entry in the docs, and NOTHING on the web. 

from `npm help scripts`

> If npm was invoked with root privileges, then it will change the uid to the user account  or uid specified by the `user` config, which defaults to `nobody`.  Set the `unsafe-perm` flag to run scripts with root privileges.

I guess that "setting the flag" is done with `--unsafe-perm` on the command line. If you are logged in as root, it will run the *entire* install, including downline scripts that are spawned by an `npm install`ation, as the root user.

I was having a problem with `npm install sails-mongo` failing on this and `--unsafe-perm` fixed it.

**Comment by hemalchevli** [comment]
Note: the reason for using the --unsafe-perm option is that when node-gyp tries to recompile any native libraries it tries to do so as a “nobody” user and often fails to get access to certain directories. This causes alarming warnings that look like errors… but sometimes are errors. Allowing node-gyp to run as root using this flag avoids this - or rather shows up any real errors instead.

Source: http://nodered.org/docs/hardware/raspberrypi.html

**Comment by isimmons** [comment]
Can it not be made to run as the current user?

**Comment by ikitashi** [comment]
--unsafe-perm works
Ach Linux

**Comment by dasGringuen** [comment]
I had this problem because of using sudo. Works if you first change to the root user :
$ su
# npm install ....

**Comment by sokolware** [comment]
--unsafe-prem does not solve this problem on raspberry pi. Looks like check sum is not valid.

```
p verb download contents checksum {"node-v0.12.0.tar.gz":"4ec4d32fa16ad1bcdb4229dc57787f26a978bad76abe02980ce17e8cf1a08109"}
gyp verb validating download checksum for node-v0.12.0.tar.gz (4ec4d32fa16ad1bcdb4229dc57787f26a978bad76abe02980ce17e8cf1a08109 == 9700e23af4e9b3643af48cef5f2ad20a1331ff531a12154eef2bfb0bb1682e32)
gyp WARN install got an error, rolling back install
gyp verb command remove [ '0.12.0' ]
gyp verb remove using node-gyp dir: /root/.node-gyp
gyp verb remove removing target version: 0.12.0
gyp verb remove removing development files for version: 0.12.0
gyp ERR! configure error 
gyp ERR! stack Error: node-v0.12.0.tar.gz local checksum 4ec4d32fa16ad1bcdb4229dc57787f26a978bad76abe02980ce17e8cf1a08109 not match remote 9700e23af4e9b3643af48cef5f2ad20a1331ff531a12154eef2bfb0bb1682e32
gyp ERR! stack
```

**Comment by jyoungrok** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

jyr3357@naver.com 님께 보내신 메일 <Re: [node-gyp] Warning "root" does not have permission to access the dev dir (#454)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by grantcarthew** [comment]
I have this issue also. 
```
gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/0.12.7"
```
I haven't tried $su yet because the command ``` $sudo npm install ``` still works with a message "attempting to reinstall using a temporary dev dir".   It does take a long time to run though.

**Comment by archonic** [comment]
I'm also getting this with 0.12.6 on Debian. Attempting to reinstall failed due to a checksum error. With `--unsafe-perm` it failed with this:

    FATAL ERROR: node::smalloc::Alloc(v8::Handle<v8::Object>, size_t, v8::ExternalArrayType) Out Of Memory
    Aborted

On a Raspberry Pi, running as root.

**Comment by Code-Crash** [comment]
+1 

I'm getting same error while installing ishout.js . 

# npm install ishout.js

gyp WARN EACCES user "root" does not have permission to access the dev dir "/var/root/.node-gyp/0.12.7"

**Comment by bnoordhuis** [comment]
The issue here is that npm normally drops privileges when run as root (e.g. through sudo) before executing node-gyp.  As a result, node-gyp may not have access to the the directory containing the node sources.  It then logs that warning and tries to soldier on but as often as not, it'll fail.

`--unsafe-perm` runs everything as root, including all package pre- and post-install scripts.  It probably "fixes" the issue in most cases but I don't really recommend it, it's better to not run npm as root.

Closing, not a bug, just a quirk.

**Comment by chernyog** [comment]
+1
I installed GitBook also being given: gyp WARN EACCES user "root" does not have permission to access the dev dir "/Users/apple/.node-gyp/0.12.7"

**Comment by RoopeKar** [comment]
Hey,
I had this error on Ubuntu 14.04 from -> sudo npm install -g webpack-dev-server. 

Installing compiler fixed it for me:
sudo apt-get install g++

For me it turned out to be a dependency issue,
hope this helps anyone :)

**Comment by EthanMichaels** [comment]
Hi @bnoordhuis I'm getting these warnings when trying to install [pm2](https://www.npmjs.com/package/pm2) on Mac OS 10.10.5.

    $ sudo npm i -g pm2

    > fsevents@0.3.8 install /usr/local/lib/node_modules/pm2/node_modules/chokidar/node_modules/fsevents
    > node-gyp rebuild

    gyp WARN EACCES user "root" does not have permission to access the dev dir "/Users/homedir/.node-gyp/0.12.7"
    gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/pm2/node_modules/chokidar/node_modules/fsevents/.node-gyp"
      SOLINK_MODULE(target) Release/.node
      CXX(target) Release/obj.target/fse/fsevents.o
      SOLINK_MODULE(target) Release/fse.node
    /usr/local/bin/pm2 -> /usr/local/lib/node_modules/pm2/bin/pm2
    /usr/local/bin/pm2-dev -> /usr/local/lib/node_modules/pm2/bin/pm2-dev

(I've also noticed similar warnings on several other `npm install`s recently.)

I understand as you said you don't consider this a bug.

You mentioned that you don't recommend using the `--unsafe-perm` option.

Do you have a recommended workaround or course of action we could take to address this?

**Comment by bnoordhuis** [comment]
> Do you have a recommended workaround or course of action we could take to address this?

I'd install it somewhere under my home directory (or _a_ home directory), not site-wide.  That way you don't have to run npm as root.

Alternatively, you can take the homebrew approach where you chown everything under /usr/local.  It's not the approach I personally would take but if you are on a single-user system (and a Mac usually is), it's probably alright.

**Comment by EthanMichaels** [comment]
> I'd install it somewhere under my home directory (or a home directory), not site-wide. That way you don't have to run npm as root.
>
> Alternatively, you can take the homebrew approach where you chown everything under /usr/local. It's not the approach I personally would take but if you are on a single-user system (and a Mac usually is), it's probably alright.

Ok, thanks for the advice. That's helpful.

**Comment by seonixx** [comment]
unsafe_perm worked for me also, however I still don't understand why it's necessary. Not really a fix, just a dirty workaround.

**Comment by aliquid** [comment]
Alternative to unsafe_perm (after install failed);
```
cd /usr/lib/node_modules # or /usr/local/lib/node_modules
ls -l                    # look for owner nobody (or any non-root owner)
chown -R root xyzzy1     # replace xyzzy with whatever from above
cd xyzzy1                # replace xyzzy with whatever from above
npm install              # will rebuild the needful
cd ../xyzzyN             # rinse and repeat
```

**Comment by aliquid** [comment]
"ls -l /usr/lib/node_modules" (or /usr/local/lib/node_modules) shows that during install the module directory switches from root owner to nobody owner. Safety precaution when compiling native binaries? Don't know. unsafe_perm may well be what is needed, I got it installed by "chmod -R root ..".

**Comment by raphaelparent** [comment]
On OS X I simply installed but didn't run it, running it for the first time and accepting the license fixed the problem for me.

**Comment by bgSosh** [comment]
+1 (Debian 7)

**Comment by SerkanSipahi** [comment]
+1 @BaggersIO , thank you, works well :)

**Comment by VitoVan** [comment]
For me:
```rm -rf /root/.node-gyp/```
```npm install --unsafe-perm --verbose -g <package-name>```
Solved my problem.

And before that, I deleted ```node_modules``` and reinstalled Node.js.

**Comment by jbburf** [comment]
I'm getting the same thing, I just installed NodeJS 4.1.2 from https://nodejs.org/en/ and get this when I run:

```
sudo npm install sails -global
```

**Comment by SerkanSipahi** [comment]
@jbburf use `npm install --unsafe-perm --verbose -g <package-name>`. In most cases it helps !

**Comment by iSynth** [comment]
a ~ $ sudo npm install -g --unsafe-perm devtools-terminal                                                                                                                                                                        
/                                                                                                                                                                                                                                
> pty.js@0.3.0 install /usr/lib/node_modules/devtools-terminal/node_modules/pty.js                                                                                                                                               
> node-gyp rebuild                                                                                                                                                                                                               
                                                                                                                                                                                                                                 
make: Entering directory `/usr/lib/node_modules/devtools-terminal/node_modules/pty.js/build'                                                                                                                                     
  CXX(target) Release/obj.target/pty/src/unix/pty.o                                                                                                                                                                              
make: g++: Command not found                                                                                                                                                                                                     
make: *** [Release/obj.target/pty/src/unix/pty.o] Error 127
make: Leaving directory `/usr/lib/node_modules/devtools-terminal/node_modules/pty.js/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System Linux 3.19.0-33-generic
gyp ERR! command "node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /usr/lib/node_modules/devtools-terminal/node_modules/pty.js
gyp ERR! node -v v0.10.40
gyp ERR! node-gyp -v v1.0.1
gyp ERR! not ok 

> ws@0.4.32 install /usr/lib/node_modules/devtools-terminal/node_modules/socket.io/node_modules/socket.io-client/node_modules/ws
> (node-gyp rebuild 2> builderror.log) || (exit 0)

make: Entering directory `/usr/lib/node_modules/devtools-terminal/node_modules/socket.io/node_modules/socket.io-client/node_modules/ws/build'
  CXX(target) Release/obj.target/bufferutil/src/bufferutil.o
make: Leaving directory `/usr/lib/node_modules/devtools-terminal/node_modules/socket.io/node_modules/socket.io-client/node_modules/ws/build'
npm ERR! pty.js@0.3.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pty.js@0.3.0 install script.
npm ERR! This is most likely a problem with the pty.js package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls pty.js
npm ERR! There is likely additional logging output above.

npm ERR! System Linux 3.19.0-33-generic
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "install" "-g" "--unsafe-perm" "devtools-terminal"
npm ERR! cwd /home/a
npm ERR! node -v v0.10.40
npm ERR! npm -v 1.4.28
npm ERR! code ELIFECYCLE
npm ERR! not ok code 0
a ~ $

**Comment by jyoungrok** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

jyr3357@naver.com 님께 보내신 메일 <Re: [node-gyp] Warning "root" does not have permission to access the dev dir (#454)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by grantcarthew** [comment]
Hi everyone,
This guide fixed the problem for me perfectly.
https://github.com/sindresorhus/guides/blob/master/npm-global-without-sudo.md

I will copy the text here for future reference

# Install `npm` packages globally without sudo on OS X and Linux

`npm` installs packages locally within your projects by default. You can also install packages globally (e.g. `npm install -g <package>`) (useful for command-line apps). However the downside of this is that you need to be root (or use `sudo`) to be able to install globally.

Here is a way to install packages globally for a given user.

###### 1. Create a directory for your global packages

```sh
mkdir "${HOME}/.npm-packages"
```

###### 2. Reference this directory for future usage in your `.bashrc`/`.zshrc`:

```sh
NPM_PACKAGES="${HOME}/.npm-packages"
```

###### 3. Indicate to `npm` where to store your globally installed package. In your `$HOME/.npmrc` file add:

```sh
prefix=${HOME}/.npm-packages
```

###### 4. Ensure `node` will find them. Add the following to your `.bashrc`/`.zshrc`:

```sh
NODE_PATH="$NPM_PACKAGES/lib/node_modules:$NODE_PATH"
```

###### 5. Ensure you'll find installed binaries and man pages. Add the following to your `.bashrc`/`.zshrc`:

```sh
PATH="$NPM_PACKAGES/bin:$PATH"
# Unset manpath so we can inherit from /etc/manpath via the `manpath`
# command
unset MANPATH # delete if you already modified MANPATH elsewhere in your config
MANPATH="$NPM_PACKAGES/share/man:$(manpath)"
```

-

Check out [`npm-g_nosudo`](https://github.com/glenpike/npm-g_nosudo) for doing the above steps automagically

-

NOTE: If you are running OS X, the `.bashrc` file may not yet exist, and the terminal will be obtaining its environment parameters from another file, such as `.profile` or `.bash_profile`. These files also reside in the user's home folder. In this case, simply adding the following line to them will instruct Terminal to also load the `.bashrc` file:

```sh
source ~/.bashrc
```

---

See also: `npm`'s documentation on
["Fixing `npm` permissions"](https://docs.npmjs.com/getting-started/fixing-npm-permissions).

**Comment by willxy** [comment]
Thanks for the explanation about root & nobody; that wasn't clear in the output and it explains the problem.  I worked around it by temporarily clowning the (in my case) $HOME/.node-gyp/6.1.0 to nobody, and that let the install/compile proceed.

I agree that running with --unsafe-perm is not advisable, and thus I would not consider that a valid workaround.

It seems like, if it's going to drop privs to "nobody" before continuing, it should probably, before it drops privs, set up a directory that it knows it will be able to write to -- rather than dropping privs and then trying to write to a directory it almost certainly won't be able to write to as "nobody" (a dir under $HOME or /usr/local, in my case).

Could it not "mkdir /tmp/whatever; chown nobody /tmp/whatever" and then proceed with the build using that directory?

**Comment by tbaltrushaitis** [comment]
Hi there. I've solved those terrible issue for my 
```
# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
```
with this:
```
yum groupinstall 'Development Tools'
npm i --unsafe-perm
```
Hope this helps.

**Comment by Alblahm** [comment]
Hey, I got a similar problem as yours, I missed homebridge node. 

I realize that it was in some way related with some cache problem. I tested a lot of options but at the end the solution was: Enter in the BridgeÇore.js file, and change the username address of the Bridge node.
 username: "CC:22:3D:E3:CE:F6", change this and reload ...

I suppose that it is because, it is a new id, it means a new bridge. I'm completely sure that it was the problem, because if I undo the username and revert to the original value, it does not work.

**Comment by DanielRuf** [comment]
It seems this is still not fixed =(

**Comment by gibfahn** [comment]
>It seems this is still not fixed =(

"Don't install node/npm as root" is the fix. The issue you're seeing (see https://github.com/nodejs/node-gyp/issues/454#issuecomment-134140378) is that npm doesn't like to run as root, if you want to change that then feel free to file an issue with npm (or just search their issues, I'm sure lots of people have complained before): https://github.com/npm/npm/issues

If you're seeing this without having used `sudo` then let us know.

**Comment by DanielRuf** [comment]
I'm not sure why this should be the right solution.

Also see https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions

You do not come far without sudo ;-)

**Comment by bnoordhuis** [comment]
Installing node or npm with sudo, sure.  But running them as root?  Should never be necessary.

**Comment by DanielRuf** [comment]
> Installing node or npm with sudo, sure. But running them as root? Should never be necessary.

Guess what, the problem occurs when I install without sudo (I did not install the package with sudo). The `unsafe-perm` flag was the only reliable solution here. ;-)

**Comment by DanielRuf** [comment]
https://github.com/nodejs/node-gyp/issues/454#issuecomment-48204995
https://github.com/nodejs/node-gyp/issues/454#issuecomment-58792114
https://github.com/nodejs/node-gyp/issues/454#issuecomment-134140378
https://github.com/nodejs/node-gyp/issues/454#issuecomment-251272404

The problem occurs with node 6 and npm 3 on a Debian 8 root server when I try to install sitespeed.io. With a normal `npm install -g sitespeed.io` command. Nothing special on my side.

**Comment by bnoordhuis** [comment]
Let me rephrase: if it doesn't work without root, chances are you are doing something wrong.

If `npm install -g` in your case means 'install to /usr/local', ask yourself why /usr/local and not your home directory.  Node, npm and node-gyp don't mandate that packages are installed system-wide.

If sitespeed.io does, that's a bug in that package, but you are in the wrong bug tracker in that case.

**Comment by gibfahn** [comment]
@DanielRuf see the npm help pages, which go into this quite exhaustively: https://docs.npmjs.com/getting-started/fixing-npm-permissions

**Comment by DanielRuf** [comment]
Well, I tried option 2 with the custom path / directory and it did not fix the issue. And I think I also tried option 1 too but npm always reset the directory. Anyway, I got it working with the `unsafe-perm` flag.

**Comment by gibfahn** [comment]
>Well, I tried option 2 with the custom path / directory and it did not fix the issue.

A global modules directory that is owned by the user (not by root) is the default for every install method that isn't a Linux package manager, and it works fine.

>npm always reset the directory

Unless you're still running with `sudo npm` then I've never known it do that.

<br/>

For anyone else coming across this thread, I recommend just using `nvm`, which avoids these issues. Using `--unsafe-perm` (running random module install scripts as root) is a pretty terrible idea.

**Comment by tbaltrushaitis** [comment]
> I recommend just using nvm

:one: Install with a [line of code](https://github.com/tbaltrushaitis/tower-strike#install-nvm)
:two: Use
```bash
$ nvm ls
$ nvm ls-remote
$ nvm install
$ nvm use
$ nvm exec
```
:white_check_mark: Enjoy [your life](https://github.com/tbaltrushaitis/tower-strike#usage) ;-)

**Comment by e-beach** [comment]
The configuration of global packages should be up to the user, and npm defaults to `/usr/local`. This package should work with those settings, and not run into an *infinite loop* of `attempting to reinstall using temporary dev dir "/usr/lib/node_modules/node-inspector/node_modules/v8-profiler/.node-gyp"`.

**Comment by bnoordhuis** [comment]
@e-beach Have you actually read through the issue or is yours just a pile-up comment?

**Comment by ORESoftware** [comment]
Note that --unsafe-perm has to come before the package name:

FAIL

`RUN sudo npm install -g "github:sumanjs/suman#rebase_branch" --only=production --unsafe-perm`

SUCCESS:

`RUN sudo npm install --only=production --unsafe-perm -g "github:sumanjs/suman#rebase_branch"`

**Comment by Davenchy** [comment]
Try
`sudo npm install -g PackageName --allow-root`

**Comment by JSAbrahams** [comment]
@Davenchy Nope didn't work.

**Comment by jasmeetdd** [comment]
@Davenchy 
Thanks buddy!
This:-
`sudo npm install -g PackageName --allow-root`
Dis the trick for me on CentOS with cPanel.

**Comment by Extarys** [comment]
Same issue here with angular. 

`sudo npm i -g @angular/cli --allow-root` Did the trick.

Never had issue with npm and node before and now with clean os I had so many issues in the last two hours I cant believe it.

**Comment by manmohanbishnoi** [comment]
Issue on Ubuntu 16.04.2 LTS when using
`sudo npm install -g @angular/cli --allow-root`

This one fixed the errors and installation completed successfully
`sudo npm install --unsafe-perm -g @angular/cli`

**Comment by bnoordhuis** [comment]
I've taken the liberty of locking this because it's turning into one of those undead issues that attract comments long after its expiry date.

For people coming here through search engines, see https://github.com/nodejs/node-gyp/issues/454#issuecomment-315691803.
